src/rgw/rgw_rados.cc

```diff
@@ -7052,25 +7046,75 @@ int RGWRados::olh_init_modification(const DoutPrefixProvider *dpp, const RGWBuck
   return ret;
 }
 
-int RGWRados::guard_reshard(const DoutPrefixProvider *dpp,
-                            BucketShard *bs,
-			    const rgw_obj& obj_instance,
-			    const RGWBucketInfo& bucket_info,
-			    std::function<int(BucketShard *)> call)
+int RGWRados::guard_reshard_read(const DoutPrefixProvider *dpp,
+                                 const rgw_obj& obj_instance,
+                                 const RGWBucketInfo& bucket_info,
+                                 std::function<int(BucketShard *)> call)
 {
   rgw_obj obj;
   const rgw_obj *pobj = &obj_instance;
   int r;
+  BucketShard bs(this);
 
-  for (int i = 0; i < NUM_RESHARD_RETRIES; ++i) {
-    r = bs->init(pobj->bucket, *pobj, nullptr /* no RGWBucketInfo */, dpp);
-    if (r < 0) {
-      ldpp_dout(dpp, 5) << "bs.init() returned ret=" << r << dendl;
+  // one retry is enough here because once we fetch the new bucket ID,
+  // we shouldn't get another ENOENT due to resharding
+  for (int i = 0; i < 2; ++i) {
+    r = bs.init(pobj->bucket, *pobj, nullptr /* no RGWBucketInfo */, dpp);
+    if (r != -ENOENT) {
+      if (r < 0) {
+        ldpp_dout(dpp, 5) << "bs.init() returned ret=" << r << dendl;
+        return r;
+      }
+      r = call(&bs);
+      if (r != -ENOENT) {
+        return r;
+      }
```

Comment:
`guard_reshard_read()` is being called from `bi_get()` to read a specific key from the bucket index. `cls_rgw_bi_get()` returns `-ENOENT` if that key isn't found, which `guard_reshard_read()` is misinterpreting to mean that the bucket index shard object itself wasn't found

because of the structure of this for-loop, even though `call()` returns errors both times, the later call to `fetch_new_bucket_id()` succeeds so `guard_reshard_read()` ends up returning its successful error code. that's why `bi_get_instance()` goes on to decode an empty buffer and fail with `ERROR: failed to decode bi_entry()`

i don't think we ever hit this case with `guard_reshard()` on writes, because writes to nonexistent keys would create them rather than fail with `ENOENT`

it seems we need a way to differentiate between errors from the cls call and the osd op as a whole. although the `IoCtx::exec()` function used by `cls_rgw_bi_get()` doesn't give back the error code from cls, there are overloads of `ObjectOperation::exec()` that do - both the one taking `int *prval` and `ObjectOperationCompletion*` should have access to the error code from the cls call. if that value comes back `-ENOENT`, `guard_reshard_read()` should return it directly instead of treating it like a reshard event. but if the `IoCtx::operate()` returns `-ENOENT` and `ObjectOperation::exec()` doesn't, that should imply that the shard object is gone

i'm not sure how to handle this correctly without making a mess of the `cls_rgw_bi_get()` and `guard_reshard_read()` interfaces. what do you think?

Reply:
Hey @cbodley, thanks for taking the time to dig into those issues, that analysis was very helpful. At this point, this fix is not as critical from my standpoint. The other patches that we've merged ensure that the OLH object doesn't get corrupted and that extra index entries aren't generally left behind in this scenario. The fact that PUT requests during a reshard fail with a 404 error still isn't ideal, but reshards usually take at least a few minutes and these requests almost always fail for us regardless due to gateway or client timeouts. I'm satisfied with the fact that this is at least resolved in Reef. Maybe in Quincy/Pacific we should just consider changing the status code in this scenario to make it less confusing - maybe a 503?

Reply:
> At this point, this fix is not as critical from my standpoint.

👍 

> Maybe in Quincy/Pacific we should just consider changing the status code in this scenario to make it less confusing - maybe a 503?

it definitely makes sense to return a retryable http error. would `guard_reshard()` do this mapping from ENOENT to something else like ERR_RATE_LIMITED or ERR_BUSY_RESHARDING?

Reply:
bump!

Reply:
any plans to pursue this, or should we close https://tracker.ceph.com/issues/61710?

Reply:
I'm gonna look at it this week and make a decision about whether to pursue further improvements or close it.